### PR TITLE
Make "make check" pass in an uninstalled --enable-mca-dso tree

### PR DESCRIPTION
**Pull request by jsquyres**
### What

Make `make check` succeed in an uninstalled `--enable-mca-dso` build tree.

### Why

With `--enable-mca-dso`, MCA components are built as standalone DSOs that
are `dlopen`'ed at run time from the install tree (`$libdir/openmpi`).
But `make check` runs before `make install`, so at test time those DSOs
exist only in the build tree under libtool `.libs` directories.  Tests
that call `opal_init()` / `MPI_Init()` therefore fail — for example,
`opal_init()` aborts in `opal_shmem_base_select()` because no shmem
component can be loaded.

In an uninstalled DSO build this currently breaks:

- `test/threads/opal_condition`
- `test/datatype/mpi_datatype_bigcount`
- all of `test/mpi/file/*`

### How

A shared Automake fragment, `test/Makefile.mca-dso-check`, `include`d by
the affected test directories, fixes this in two ways:

1. Point the Open MPI and PMIx MCA component search paths at the
   build-tree `.libs` directories holding the freshly-built component
   DSOs (computed at test-run time), so `opal_init()` / `MPI_Init()` can
   find them.  The exports are guarded so an empty result never clobbers
   a default search path (e.g. an external PMIx).

2. Symlink each project's `mca_common` shared libraries into that
   project's own `.libs` directory.  Some component DSOs carry a
   load-time dependency on an `mca_common` library shared by a family of
   sibling components (e.g. `io/ompio`, `fcoll/*`, `fbtl/*`, and
   `sharedfp/*` all depend on `libmca_common_ompio`).  Those common
   libraries live under `<project>/mca/common/<name>/.libs`, which is not
   on the runtime library search path baked into the libtool test
   wrapper scripts — the wrapper only lists `<project>/.libs`.  On macOS
   this cannot be worked around via `DYLD_LIBRARY_PATH` in
   `AM_TESTS_ENVIRONMENT`, because System Integrity Protection strips
   `DYLD_*` when the `/bin/sh` wrapper is exec'ed; symlinking into
   `<project>/.libs`, which the wrapper already searches, works on both
   macOS and Linux.

Both mechanisms are harmless for a static (non-DSO) build, and the
symlinks live in build-tree `.libs` directories that are removed
automatically by libtool's existing `clean-libtool` rule.

### Testing

Verified on macOS:

- `make check` in an uninstalled `--enable-mca-dso` tree — all test
  directories pass.
- `make distcheck` with `--enable-mca-dso` — succeeds (clean-room VPATH
  build → `make check` → install → distclean leftover-file check all
  pass, confirming the symlinks are cleaned up and do not break
  `distcheck`).
- `make check` in a static (non-DSO) build — all test directories pass;
  the fragment is a no-op there (no symlinks created, component-path
  exports harmless).

### CODE REVIEW

**Review by hppritcha — DISMISSED** [review]

File: test/Makefile.mca-dso-check [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#
+# NOTE: this fragment defines an "all-local" rule.  Automake forbids two
+# definitions of the same *-local target in one Makefile.am, so a
+# Makefile.am that includes this fragment must NOT define its own
+# "all-local"; instead, add prerequisites to the mca-common-dso-symlinks
+# target below.
+
+AM_TESTS_ENVIRONMENT = \
+	ompi_mca_path="`find $(abs_top_builddir)/opal/mca $(abs_top_builddir)/ompi/mca $(abs_top_builddir)/oshmem/mca -type d -name .libs 2>/dev/null | sort | tr '\n' ':'`"; \
+	pmix_mca_path="`find $(abs_top_builddir)/3rd-party/openpmix/src/mca -type d -name .libs 2>/dev/null | sort | tr '\n' ':'`"; \
```
Comment (bwbarrett) [inline test/Makefile.mca-dso-check]: it's a corner case, but I'd feel a little bit better if we only did the PMIx stuff when we were using an internal build of PMIx.  the use case I'm thinking of is:

* build OMPI with internal pmix
* rebuild OMPI, but this time with external pmix
* run "make check", and we're going to configure PMIx to find the components from step 1, but use the library from step 2.

**Review by bwbarrett — DISMISSED** [review]

File: test/Makefile.mca-dso-check (hunk 1 above)
Comment (jsquyres) [reply test/Makefile.mca-dso-check]: Good catch — fixed. The PMIx search-path logic is now gated on `@OPAL_USING_INTERNAL_PMIX@`: the `pmix_mca_path` `find` under `3rd-party/openpmix` and the `PMIX_MCA_mca_base_component_path` export only run when Open MPI was built against its internal PMIx. With an external PMIx (including your rebuild-with-external-after-internal case) that block is skipped entirely, so a stale internal-PMIx build in the tree can't be pointed at by a run that links the external PMIx library. Verified the substitution: internal builds expand to `if test "1" = "1"`, external to `if test "0" = "1"`. Amended into the single commit.

**Review by bwbarrett — APPROVED** [review]
